p2p/pex/reactor.go

```diff
@@ -170,56 +359,157 @@ func (r *ReactorV2) handleMessage(chID p2p.ChannelID, envelope p2p.Envelope) (er
+	switch peerUpdate.Status {
+	case p2p.PeerStatusUp:
+		r.availablePeers.PushBack(peerUpdate.NodeID)
+	case p2p.PeerStatusDown:
+		r.removePeer(peerUpdate.NodeID)
+	default:
+	}
+}
 
-	for {
-		select {
-		case envelope := <-r.pexCh.In:
-			if err := r.handleMessage(r.pexCh.ID, envelope); err != nil {
-				r.Logger.Error("failed to process message", "ch_id", r.pexCh.ID, "envelope", envelope, "err", err)
-				r.pexCh.Error <- p2p.PeerError{
-					NodeID: envelope.From,
-					Err:    err,
-				}
-			}
+func (r *ReactorV2) waitUntilNextRequest() <-chan time.Time {
+	return time.After(time.Until(r.nextRequestTime))
+}
 
-		case <-r.closeCh:
-			r.Logger.Debug("stopped listening on PEX channel; closing...")
-			return
+// sendRequestForPeers pops the first peerID off the list and sends the
+// peer a request for more peer addresses. The function then moves the
+// peer into the requestsSent bucket and calculates when the next request
+// time should be
+func (r *ReactorV2) sendRequestForPeers() {
+	peer := r.availablePeers.Front()
+	if peer == nil {
+		// no peers are available
+		r.Logger.Debug("no available peers to send request to, waiting...")
+		r.nextRequestTime = time.Now().Add(noAvailablePeersWaitPeriod)
+		return
+	}
+	peerID := peer.Value.(p2p.NodeID)
+
+	// The node accommodates for both pex systems
+	if r.isLegacyPeer(peerID) {
+		r.pexCh.Out <- p2p.Envelope{
+			To:      peerID,
+			Message: &protop2p.PexRequest{},
+		}
+	} else {
+		r.pexCh.Out <- p2p.Envelope{
+			To:      peerID,
+			Message: &protop2p.PexRequestV2{},
 		}
 	}
+
+	// remove the peer from the available peers list and mark it in the requestsSent map
+	r.availablePeers.Remove(peer)
+	peer.DetachPrev()
+	r.mtx.Lock()
+	r.requestsSent[peerID] = struct{}{}
+	r.mtx.Unlock()
+
+	r.calculateNextRequestTime()
+	r.Logger.Debug("peer request sent", "next_request_time", r.nextRequestTime)
 }
 
-// processPeerUpdate processes a PeerUpdate. For added peers, PeerStatusUp, we
-// send a request for addresses.
-func (r *ReactorV2) processPeerUpdate(peerUpdate p2p.PeerUpdate) {
-	r.Logger.Debug("received peer update", "peer", peerUpdate.NodeID, "status", peerUpdate.Status)
+// calculateNextRequestTime implements something of a proportional controller
+// to estimate how often the reactor should be requesting new peer addresses.
+// The dependent variable in this calculation is the ratio of new peers to
+// all peers that the reactor receives. The interval is thus calculated as the
+// inverse squared. In the beginning, all peers should be new peers.
+// We  expect this ratio to be near 1 and thus the interval to be as short
+// as possible. As the node becomes more familiar with the network the ratio of
+// new nodes will plummet to a very small number, meaning the interval expands
+// to its upper bound.
+// MaxInterval = 100 * 100 * baseTime ~= 16mins for 10 peers
+// CONTRACT: Must use a write lock as nextRequestTime is updated
+func (r *ReactorV2) calculateNextRequestTime() {
```

Comment:
I haven't fully reviewed this in detail but I trust your judgement here 👍

Reply:
Thanks. I'm hoping I've tested/thought about the heuristics here sufficiently for this to be solid.